lab/compute: Replace D code with C in `sum-array` (#270)
open-education-hub/operating-systems: content/chapters/compute/lab/content/arena.md

## Libraries for Parallel Processing

In `support/sum-array/c/sum_array_threads.c` we spawned threads "manually" by using the `spawn` function.

> This is **not** a syscall, but a wrapper over the most common thread-management API in POSIX-based operating systems (such as Linux, FreeBSD, macOS): POSIX Threads or `pthreads`.
> By inspecting the [implementation of `spawn`](https://github.com/dlang/phobos/blob/352258539ca54e640e862f79b2b8ec18aafa7d94/std/concurrency.d#L618-L622), we see that it creates a `Thread` object, on which it calls the `start()` method.
> In turn, [`start()` uses `pthread_create()`](https://github.com/dlang/dmd/blob/cc117cd45c7d72ce5a87b775e65a9d13fa4d4424/druntime/src/core/thread/osthread.d#L454-L486) on POSIX systems.

Reviewer: ```suggestion
This is **not** a syscall, but a wrapper over the common syscall used by both `fork()` (which is also not a syscall) and `pthread_create()`.
```

In order to see what syscall `pthread_create()` uses, check out [this section at the end of the lab](#threads-and-processes-clone).

Most programming languages provide a more advanced API for handling parallel computation.

> C does not, by design.
> It needs a library to do advanced things, like `reduce` from D, an example of which can be found in `support/sum-array/d/sum_array_threads_reduce.d`.

Reviewer: ```suggestion
### `std.parallelism` in D
(add the previous text here)

### OpenMP for C
Unlike D, C does not support parallel computation by design.
It needs a library to do advanced things, like `reduce()` from D.
```

We have chosen to use the OpenMP library for this.
Follow the code in `support/sum-array/c/sum_array_threads_openmp.c`.

The `#pragma` used in the code instructs the compiler to enable the `omp` module, and to parallelise the code.

> In this case, we instruct the compiler to perform a reduce of the array, using the `+` operator, and to store the results in the `result` variable.

Reviewer: ```suggestion
In this case, we instruct the compiler to perform a reduce of the array, using the `+` operator, and to store the results in the `result` variable.
This reduction uses threads to calculate the sum, similar to `summ_array_threads.c`, but in a much more optimised form.
```

In order to see what syscall `pthread_create()` uses, check out [this section at the end of the lab](#threads-and-processes-clone).

Most programming languages provide a more advanced API for handling parallel computation.

### `std.parallelism` in D

D language's standard library exposes the [`std.parallelism`](https://dlang.org/phobos/std_parallelism.html), which provides a series of parallel processing functions.

> One such function is `reduce` which splits an array between a given number of threads and applies a given operation to these chunks.

Reviewer: ```suggestion
One such function is `reduce()` which splits an array between a given number of threads and applies a given operation to these chunks.
```

This structure is a thread manager (not scheduler).

> It silently creates the number of threads we request and then `reduce` spreads its workload between these threads.

Reviewer: ```suggestion
It silently creates the number of threads we request and then `reduce()` spreads its workload between these threads.
```
